rfcs/20231017-collective-broadcast.md

```diff
@@ -0,0 +1,94 @@
+# [RFC] Add collective_broadcast to the StableHLO specification
+
+Status: Review<br/>
+Initial version: 10/17/20223<br/>
+Last updated: 10/17/2023<br/>
+Discussion thread: [GitHub](https://github.com/openxla/stablehlo/pull/1809)
+
+## Motivation
+
+StableHLO currently has [five collective communication primitives](https://github.com/openxla/stablehlo/blob/main/docs/spec.md#collective-ops): `collective_permute`, `all_gather`, `all_to_all`, `all_reduce`, and `reduce_scatter`. However, one of the major collective communication primitives, `broadcast`, is missing from this list. This primitive allows for a one-to-many replication of a tensor to many devices efficiently. `broadcast` is a primitive in [MPI](https://www.open-mpi.org/doc/v4.1/man3/MPI_Bcast.3.php), [NCCL](https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/api/colls.html#c.ncclBroadcast), and [PyTorch](https://pytorch.org/docs/stable/distributed.html#torch.distributed.broadcast). From here on out, we will refer to this operation as `collective_broadcast` for reasons discussed later.
+
+While it technically would be possible to replicate a broadcast with a conditional mask and a `psum`, that reduces to an `all_reduce` communication primitive, which is significantly more expensive than a simple `collective_broadcast`. Additionally, when dealing with network-switch environments, the explicit use of `collective_broadcast` allows the switch to greatly optimize it's throughput when replicating to many targets simultaneously. However, XLA currently has no ability to lower directly to a mesh's `collective_broadcast` primitive, so a lot of that optimization is left on the table. 
```

Comment:
> However, XLA currently has no ability to lower directly to a mesh's `collective_broadcast` primitive, so a lot of that optimization is left on the table.

IIUC this new op is only useful if a dedicated HLO is added for it in [openxla/xla](https://github.com/openxla/xla). Prior to adding to StableHLO, would be good to get buy-in / agreement on adding a new HLO.

Reply:
Sure, is there a similar RFC mechanism to raise to raise awareness to XLA proper?

Reply:
I believe sufficient notice has been given on email chains / chat threads / etc. HLO internal/external stakeholders are aware that this RFC also applies to a new HLO opcode. I will also make sure that the proper reviewers are tagged on the PR which implements the HLO.